### PR TITLE
tests: add nightly test for Protocol Arbitrary impl to assert it covers all the Protocol variants.

### LINKED ISSUE #65: tests/lib: Update `impl Arbitrary for Proto`
`Arbitrary` implementation seems to be missing some enum variants.

`Protocol` has `30` variants. `Arbitrary` implementation generates `25`.

Ideally this would be fixed by removing the need to list every `Protocol` variant. Though I can not think of a clean way of doing that. Simply updating the `Arbitrary` implementation seems fine for now.

https://github.com/multiformats/rust-multiaddr/blob/1cfb9239d7447285bfb0f329320945a5e8afabe8/tests/lib.rs#L88-L139

### PR DESCRIPTION
**Pull request by jxs**
This PR follows both @mxinden and @thomaseizinger on #65. 
Tried to add multiple toolchains in a single step [but apparently it's not possible](https://github.com/actions-rs/toolchain/issues/215).

Test running [here](https://github.com/multiformats/rust-multiaddr/actions/runs/3743596561/jobs/6355945270#step:10:69)

Closes #65

### CODE REVIEW

File: tests/lib.rs [hunk 1]
```diff
@@ -639,3 +641,12 @@ fn protocol_stack() {
         assert_eq!(ps, toks);
     }
 }
+
+// Assert all `Protocol` variants are covered
+// in its `Arbitrary` impl.
+#[cfg(nightly)]
```
Comment (thomaseizinger) [inline tests/lib.rs:647]: Nicely done! At first I thought that we'd need the rust-version proc macro here but `cfg`s also work very well!

File: tests/lib.rs [hunk 2]
```diff
@@ -12,6 +12,8 @@ use std::{
     str::FromStr,
 };
 
+const IMPL_VARIANT_COUNT: u8 = 32;
```
Comment (thomaseizinger) [inline tests/lib.rs:15]: We could make this a const on the type :)

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -54,6 +59,16 @@ jobs:
...
           args: --all-features --workspace
 
+      - name: Nightly Test
+        uses: actions-rs/cargo@v1
+        env:
+          RUSTFLAGS: '--cfg nightly -Zcrate-attr=feature(variant_count)'
+        with:
+          toolchain: nightly
+          command: test
+          args: nightly
```
Comment (thomaseizinger) [inline .github/workflows/build.yml:69]: What does this do?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -54,6 +59,16 @@ jobs:
           command: test
           args: --all-features --workspace
 
+      - name: Nightly Test
+        uses: actions-rs/cargo@v1
+        env:
+          RUSTFLAGS: '--cfg nightly -Zcrate-attr=feature(variant_count)'
```
Comment (thomaseizinger) [inline .github/workflows/build.yml:65]: I forgot that we need to activate features, this is 🔥🔥🔥

**Review by thomaseizinger — APPROVED** [review]
Very nicely done, I am surprised how clean this came out!

File: .github/workflows/build.yml (hunk 3 above)
Comment (jxs) [reply .github/workflows/build.yml:69]: you mean the `nightly` args? It's just so we filter and only run the tests with `nightly` on the name. I tried to think of a way to use the `#[cfg(nightly)]` but couldn't come with a way.

File: tests/lib.rs (hunk 2 above)
Comment (jxs) [reply tests/lib.rs]: ehe I wasn't even aware of that feature, thanks Thomas! :heart:

File: .github/workflows/build.yml (hunk 3 above)
Comment (thomaseizinger) [reply .github/workflows/build.yml:69]: Ah right. Can't we just run all tests? What is the issue with that? :)

File: tests/lib.rs [hunk 5]
```diff
@@ -85,10 +85,14 @@ impl Arbitrary for Ma {
 #[derive(PartialEq, Eq, Clone, Debug)]
 struct Proto(Protocol<'static>);
 
+impl Proto {
+    const IMPL_VARIANT_COUNT: u8 = 32;
```
Comment (thomaseizinger) [inline tests/lib.rs:89]: I was actually thinking of directly on the enum as `Protocol::VARIANT_COUNT`.

I am sure there are also proc-macros for this that we could enable only during testing if this nightly stuff is too fancy 😁

File: .github/workflows/build.yml (hunk 3 above)
Comment (jxs) [reply .github/workflows/build.yml:69]: Just redundancy, do you suggest then that we only run tests once and with nightly?

File: tests/lib.rs (hunk 5 above)
Comment (jxs) [reply tests/lib.rs:89]: but `Protocol` lives in the crate, the `VARIANT_COUNT` refers to `Proto`'s arbitrary implementation count. `Protocol`'s count is what we get with `mem::variant_count` you mean implementing it on `Protocol` in the tests?

Comment (thomaseizinger) [reply tests/lib.rs:89]: Right, that would unnecessarily increase the public API. Leave it as is! :)

File: .github/workflows/build.yml (hunk 3 above)
Comment (mxinden) [reply .github/workflows/build.yml:69]: I don't think CI time matters much in this repository. In other words, I would assume running tests twice is not an issue.

**Review by mxinden — COMMENTED** [review]
Direction looks good to me. Thanks for removing the footgun.

**Comment by jxs** [comment]
> Direction looks good to me. Thanks for removing the footgun.

ok thanks, updated!

**Review by mxinden — APPROVED** [review]

**Pull request merged**